src/14_legendre.jl

```diff
@@ -0,0 +1,372 @@
+
+    # step 2: compute P_{m+1}^m(x)
+    p_prev  = p                                     # = P_m^m(x)
+    p       = (2m+1) * x * p_prev                   # = P_{m+1}^m(x)
+
+    # step 3: compute P_n^m(x)
+    for k = m+2:n
+        p_prev, p_prev_prev = p, p_prev
+        p = (2k-1)/(k-m)*x*p_prev - (k+m-1)/(k-m)*p_prev_prev       # = P_k^m(x)
+    end                                             # on output: p = P_n^m(x)
+
+    p
+end
+
+@doc raw"""
+    legendreq(n[, m], x)
+
+Evaluate the Legendre function of second kind ``Q_n(x)`` of integer degree ``n`` at position
+``x``.
+If the order ``m`` is supplied then the associated Legendre function of second kind
+``Q_n^{(m)}(x)`` is computed.
+Their definitions are given by
+
+```math
+\begin{align}
+    Q_n^{(m)}(x)
+    &=
+    (-1)^m (1-x^2)^\frac{m}{2} \frac{\mathrm{d}^m}{\mathrm{d}x^m} \left( Q_n(x) \right)
+    \,, \quad m \in \mathbb{N}_0 \,, \; x \in (-1,1)
+    \\
+    Q_n(x)
+    &=
+    \begin{cases}
+        \frac{1}{2} \log \frac{1+x}{1-x}
+        & \text{for} \quad n = 0
+        \\
+        P_1(x) Q_0(x) - 1
+        & \text{for} \quad n = 1
+        \\
+        \frac{2n-1}{n} x Q_{n-1}(x) - \frac{n-1}{n} Q_{n-2}(x)
+        & \text{for} \quad n \geq 2
+        \,.
+    \end{cases}
+\end{align}
+```
+where ``P_1`` is the Legendre polynomial of first kind and degree one.
+The functions diverge at the interval boundaries
+``\lim_{x\rightarrow \pm 1} |Q_n(x)| = \infty``.
+
+External links: [DLMF - Legendre function](https://dlmf.nist.gov/14.7.E2),
+[DLMF - associated Legendre function](https://dlmf.nist.gov/14.7.E9),
+[Wikipedia - Legendre function](https://en.wikipedia.org/wiki/Legendre_functions#Legendre_functions_of_the_second_kind_(Qn)),
+[Wikipedia - associated Legendre function](https://en.wikipedia.org/wiki/Legendre_function#Associated_Legendre_functions_of_the_second_kind).
+"""
+function legendreq(n::Integer, x)
+    (0 > n)      && throw(DomainError(n, "must be non-negative"))
+    (1 < abs(x)) && throw(DomainError(x, "must be in the range [-1,1]"))
+
+    if x == 1
+        return Inf
+    elseif x == -1
+        return (-1)^(n+1) * Inf
+    end
+
+    Q0(y) = 0.5 * log((1+y)/(1-y))
+    Q1(y) = legendrep(1, x) * Q0(y) - 1
+
+    ABC_recurrence(n, x,
+        m->(2m-1)//m, 0, m->(m-1)//m,       # A_m, B_m, C_m
+        Q0, Q1)
+end
+function legendreq(n::Integer, m::Integer, x)
+    (n <  0) && throw(DomainError(n, "must be non-negative"))
+    (m <  0) && throw(DomainError(m, "must be non-negative"))
+    (m == 0) && return legendreq(n, x)
+
+    # x check after: Q^0_n is also implemented for x = +-1
+    (1 <= abs(x)) && throw(DomainError(x, "must be in the range (-1,1)"))
+
+    # step 1: compute Q_n^0(x)
+    q_n_0 = legendreq(n, x)
+
+    # step 2: compute Q_n^1(x)
+    Q01 = -(1-x^2)^(-0.5)                                                   # = Q_0^1(x)
+    Q11 = -sqrt(1-x^2) * (0.5*log((1+x)/(1-x)) + x/(1-x^2))                 # = Q_1^1(x)
```

Comment:
You compute `sqrt(1-x^2)` twice in two lines here…

Reply:
Note also that these expressions implicitly promote to double precision.  If `x` is single-precision don't we want to do the computation in single precision.?

Reply:
Changed in commit [0541245](https://github.com/JuliaMath/SpecialFunctions.jl/pull/175/commits/05412456f6fd6e961fd8a3130105158df93d5f20).

Variable `Q01` is used in computation of `Q11` s.t. `sqrt` is called only once.

Type stability is also enforced.